# Bank marketing analysis

## Importing necessary libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
import seaborn as sns
sns.set()

from scipy import stats
stats.chisqprob = lambda chisq, df: stats.chi2.sf(chisq, df)

## load the data

In [4]:
raw_data = pd.read_csv('Bank-data.csv')

### Cleaning the data

In [5]:
data =raw_data.copy()
data = data.drop(['Unnamed: 0'],axis = 1)
data['y'] = data['y'].map({'yes':1,'no':0})
data.head()

interest_rate  credit  march  may  previous  duration  y
0          1.334     0.0    1.0  0.0       0.0     117.0  0
1          0.767     0.0    0.0  2.0       1.0     274.0  1
2          4.858     0.0    1.0  0.0       0.0     167.0  0
3          4.120     0.0    0.0  0.0       0.0     686.0  1
4          4.856     0.0    1.0  0.0       0.0     157.0  0

In [15]:
data.describe()

interest_rate      credit       march         may    previous  \
count     518.000000  518.000000  518.000000  518.000000  518.000000   
mean        2.835776    0.034749    0.266409    0.388031    0.127413   
std         1.876903    0.183321    0.442508    0.814527    0.333758   
min         0.635000    0.000000    0.000000    0.000000    0.000000   
25%         1.042750    0.000000    0.000000    0.000000    0.000000   
50%         1.466000    0.000000    0.000000    0.000000    0.000000   
75%         4.956500    0.000000    1.000000    0.000000    0.000000   
max         4.970000    1.000000    1.000000    5.000000    1.000000   

          duration           y  
count   518.000000  518.000000  
mean    382.177606    0.500000  
std     344.295990    0.500483  
min       9.000000    0.000000  
25%     155.000000    0.000000  
50%     266.500000    0.500000  
75%     482.750000    1.000000  
max    2653.000000    1.000000

## Declare the dependent and independent variables

In [52]:
estimators = ['interest_rate','credit','march','previous','duration']
y = data['y']
x1_all = data[estimators]

## Multiple logistic regression

In [64]:
x_all = sm.add_constant(x1_all)
reg_log = sm.Logit(y,x_all)
results_log = reg_log.fit()
results_log.summary()

Optimization terminated successfully.
         Current function value: 0.336664
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                      y   No. Observations:                  518
Model:                          Logit   Df Residuals:                      512
Method:                           MLE   Df Model:                            5
Date:                Fri, 10 Jan 2025   Pseudo R-squ.:                  0.5143
Time:                        14:19:00   Log-Likelihood:                -174.39
converged:                       True   LL-Null:                       -359.05
Covariance Type:            nonrobust   LLR p-value:                 1.211e-77
=================================================================================
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
const            -0.0211      0.311     -0.068      0.946      -0.631       0.589
interest_rate    -0.8001      0.089     -8.943      0.000      -0.975      -0.625
credit            2.3585      1.088      2.169      0.030       0.227       4.490
march            -1.8322      0.330     -5.556      0.000      -2.478      -1.186
previous          1.5363      0.501      3.067      0.002       0.554       2.518
duration          0.0070      0.001      9.381      0.000       0.006       0.008
=================================================================================
"""

## Train confusion matrix and show the accuracy

In [65]:
def confusion_matrix(data,actual_values,model):
    pred_values = model.predict(data)
    bins=np.array([0,0.5,1])
    cm = np.histogram2d(actual_values, pred_values, bins=bins)[0]
    accuracy = (cm[0,0]+cm[1,1])/cm.sum()
    return cm, accuracy

In [66]:
confusion_matrix(x_all,y,results_log)

(array([[218.,  41.],
        [ 30., 229.]]),
 0.862934362934363)

## Test the model.

### load new data

In [67]:
raw_data2 = pd.read_csv('Bank-data-testing.csv')
data2 =raw_data2.copy()
data2 = data2.drop(['Unnamed: 0'], axis=1)
data2.head()

interest_rate  credit  march  may  previous  duration    y
0          1.313     0.0    1.0  0.0       0.0     487.0   no
1          4.961     0.0    0.0  0.0       0.0     132.0   no
2          4.856     0.0    1.0  0.0       0.0      92.0   no
3          4.120     0.0    0.0  0.0       0.0    1468.0  yes
4          4.963     0.0    0.0  0.0       0.0      36.0   no

In [68]:
data2['y'] = data2['y'].map({'yes':1, 'no':0})
data2.head()

interest_rate  credit  march  may  previous  duration  y
0          1.313     0.0    1.0  0.0       0.0     487.0  0
1          4.961     0.0    0.0  0.0       0.0     132.0  0
2          4.856     0.0    1.0  0.0       0.0      92.0  0
3          4.120     0.0    0.0  0.0       0.0    1468.0  1
4          4.963     0.0    0.0  0.0       0.0      36.0  0

In [69]:
data2.describe()

interest_rate      credit       march         may    previous  \
count     222.000000  222.000000  222.000000  222.000000  222.000000   
mean        2.922095    0.031532    0.274775    0.346847    0.099099   
std         1.891766    0.175144    0.447410    0.755950    0.299470   
min         0.639000    0.000000    0.000000    0.000000    0.000000   
25%         1.049250    0.000000    0.000000    0.000000    0.000000   
50%         1.714000    0.000000    0.000000    0.000000    0.000000   
75%         4.960000    0.000000    1.000000    0.000000    0.000000   
max         4.968000    1.000000    1.000000    4.000000    1.000000   

          duration          y  
count   222.000000  222.00000  
mean    398.860360    0.50000  
std     410.565798    0.50113  
min       6.000000    0.00000  
25%     144.750000    0.00000  
50%     255.500000    0.50000  
75%     525.250000    1.00000  
max    3643.000000    1.00000

## Declaring the dependent and independent variables

In [70]:
y_test = data2['y']
x1_test = data2[estimators]
x_test = sm.add_constant(x1_test)

## Test confusion matrix with its accuracy shown. the results compared to those of training confusion matrix to assess the model.. 

In [72]:
confusion_matrix(x_test, y_test, results_log)

(array([[93., 18.],
        [13., 98.]]),
 0.8603603603603603)

Looking at the test acccuracy we see a number which is a tiny but lower: 86.04%, compared to 86.29% for train accuracy. 